// google-scraper/lib/scrape.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('request');

var _request2 = _interopRequireDefault(_request);

var _cheerio = require('cheerio');

var _cheerio2 = _interopRequireDefault(_cheerio);

var _config = require('./config');

var config = _interopRequireWildcard(_config);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var GoogleScraper = function () {
  function GoogleScraper(options) {
    _classCallCheck(this, GoogleScraper);

    this.options = options;
  }

  _createClass(GoogleScraper, [{
    key: 'getHtml',
    value: function getHtml() {
      var option = {
        url: config.urlSearch(this.options.tld, this.options.language, this.options.results, this.options.keyword)
      };
      return new Promise(function (resolve, reject) {
        (0, _request2.default)(option, function (err, res, body) {
          if (err) {
            return reject(err);
          } else if (res.statusCode !== 200) {
            var error = new Error('Unexpected status code: ' + res.statusCode);
            error.res = res;
            return reject(error);
          }
          return resolve(body);
        });
      });
    }
  }, {
    key: 'extractLink',
    value: function extractLink(html) {
      // eslint-disable-line class-methods-use-this
      var arrayLinks = [];
      var $ = _cheerio2.default.load(html);
      $(config.selectorSearch).each(function (i, link) {
        var linkClean = $(link).attr('href').match('(?=http|https).*(?=&sa)');
        if (linkClean) {
          arrayLinks.push(linkClean[0]);
        }
      });
      return arrayLinks;
    }
  }, {
    key: 'getGoogleLinks',
    get: function get() {
      var _this = this;

      return new Promise(function (resolve, reject) {
        _this.getHtml().then(function (body) {
          return resolve(_this.extractLink(body));
        }).catch(function (err) {
          return reject(err);
        });
      });
    }
  }]);

  return GoogleScraper;
}();

exports.default = GoogleScraper;
module.exports = exports['default'];
